### DISCUSSION TITLE
Testing tower layer used by Hyper

### DISCUSSION THREAD

**Discussion by phughk** [hyperium/hyper#3803] [General]
Hi all,

I have a tower layer that I have configured to accept Hyper types.

```rust
impl hyper::service::Service<Request<Incoming>> for MyService {
    type Response = Response<String>;
    type Error = RapidRecastError;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn call(&self, req: Request<Incoming>) -> Self::Future {
       // ...
    }
}
```

I am now trying to test this, and it is proving difficult, since I cannot construct an `Incoming` type.

```rust
    let req = http::Request::builder()
        .method("POST")
        .uri("/test")
        .body(Incoming::from("whatever body"))
        .unwrap();
    let resp = tower.call(req).await;
```

Any pointers would be fantastic 🙏

**Comment by phughk** [comment]
Wondering if I should be using the http-body-utils crate, but not sure how that maps to `Incoming`
https://github.com/hyperium/http-body/tree/master/http-body-util/src

**Comment by seanmonstar** [comment]
The `Incoming` type is purposefully not meant to be constructable outside of hyper.

If you want to be able to test your service, without starting up a TCP server (which is what I do a lot, this works just fine), then you need to make your service generic over the request body. That way, it can accept both `Incoming` and also some mocked body in your unit tests.

It's a good point, I've updated the guides revamp issue (https://github.com/hyperium/hyper/issues/3411) to add a guide about testing your service.

**Comment by phughk** [comment]
For anyone wanting the solution to this, I have given the example from `@nylonicious` on Discord - it solves the problem perfectly.
https://github.com/rapidrecast/blog-post-snippets/tree/2b4a574401ea4f610c98b122cbcb489acdc401ba/blog-20241202-hyper-service/src
